src/hashable.ts

```diff
@@ -0,0 +1,271 @@
+      for (let i = 0; i < subEntries.length; ++i) {
+        const [existing] = subEntries[i];
+        if (existing === key) {
+          subEntries.splice(i, 1);
+          if (!subEntries.length) {
+            this.map.delete(key);
+          }
+          return true;
+        }
+      }
+      return false;
+    }
+    const hash = key.valueOf();
+    const subEntries = this.map.get(hash);
+    if (subEntries === undefined) {
+      return false;
+    }
+    for (let i = 0; i < subEntries.length; ++i) {
+      const [existing] = subEntries[i];
+      if (typeof existing !== 'object' || existing === null) {
+        continue;
+      }
+      if (existing.strictEquals(key)) {
+        subEntries.splice(i, 1);
+        if (!subEntries.length) {
+          this.map.delete(hash);
+        }
+        return true;
+      }
+    }
+    return false;
+  }
+
+  has(key: Key): boolean {
+    if (typeof key !== 'object' || key === null) {
+      const subEntries = this.map.get(key);
+      if (subEntries === undefined) {
+        return false;
+      }
+      for (const [existing] of subEntries) {
+        if (existing === key) {
+          return true;
+        }
+      }
+      return false;
+    }
+    const hash = key.valueOf();
+    const subEntries = this.map.get(hash);
+    if (subEntries === undefined) {
+      return false;
+    }
+    for (const [existing] of subEntries) {
+      if (typeof existing !== 'object' || existing === null) {
+        continue;
+      }
+      if (existing.strictEquals(key)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  get size(): number {
+    let total = 0;
+    for (const subEntries of this.map) {
+      total += subEntries.length;
+    }
+    return total;
+  }
+
+  *entries(): IterableIterator<[Key, Value]> {
+    for (const subEntries of this.map.values()) {
+      for (const entry of subEntries) {
+        yield entry;
+      }
+    }
+  }
+
+  *keys(): IterableIterator<Key> {
+    for (const subEntries of this.map.values()) {
+      for (const [key] of subEntries) {
+        yield key;
+      }
+    }
+  }
+
+  *values(): IterableIterator<Value> {
+    for (const subEntries of this.map.values()) {
+      // eslint-disable-next-line @typescript-eslint/no-unused-vars
+      for (const [_, value] of subEntries) {
+        yield value;
+      }
+    }
+  }
+
+  *[Symbol.iterator]() {
+    yield* this.entries();
+  }
+
+  get [Symbol.toStringTag]() {
+    return 'HashMap';
+  }
+
+  forEach(
+    callbackfn: (value: Value, key: Key, map: HashMap<Key, Value>) => void,
+    thisArg?: any
+  ): void {
+    callbackfn = callbackfn.bind(thisArg);
+    for (const subEntries of this.map.values()) {
+      for (const [key, value] of subEntries) {
+        callbackfn(value, key, this);
+      }
+    }
+  }
+}
+
+/**
+ * Python style set implementation using hashable values.
+ */
+export class HashSet<T extends Hashable | Primitive> implements Set<T> {
+  private hashmap: HashMap<T, undefined>;
```

Comment:
Is `undefined` really a good singleton type for non-values associated with present keys? I see you using `undefined` in `HashMap.get` (I presume in accordance with JS doing this for objects and stuff) so it’s better to use another singleton like `null` or maybe even `true` if it can be detached from the boolean type, than to ensure no `undefined` leaked somewhere.